File: frontend/src/microservices/tour-group/objects/TourGroupDetailsResponse.ts
import type { TourGroupSummaryResponse } from './TourGroupSummaryResponse'
import type { TourGroupMembershipResponse } from './TourGroupMembershipResponse'
import type { TourGroupMembershipTravelerResponse } from './TourGroupMembershipTravelerResponse'
import type { GroupPlanItemResponse } from './GroupPlanItemResponse'
import type { GroupPlanOptionResponse } from './GroupPlanOptionResponse'
import type { GroupPlanSelectionResponse } from './GroupPlanSelectionResponse'
import type { GroupSelectionOrderLinkResponse } from './GroupSelectionOrderLinkResponse'
import type { GroupSelectionOrderProjectionResponse } from './GroupSelectionOrderProjectionResponse'
import type { OrderResponse } from '@/microservices/order/objects/OrderResponse'

export type TourGroupDetailsResponse = {
  group: TourGroupSummaryResponse
  memberships: TourGroupMembershipResponse[]
  membershipTravelers: TourGroupMembershipTravelerResponse[]
  planItems: GroupPlanItemResponse[]
  planOptions: GroupPlanOptionResponse[]
  selections: GroupPlanSelectionResponse[]
  selectionOrderLinks: GroupSelectionOrderLinkResponse[]
  selectionOrderProjections: GroupSelectionOrderProjectionResponse[]
  bookings: OrderResponse[]
}
export const tourGroupDetailsResponseFromJson = (json: string): TourGroupDetailsResponse =>
  JSON.parse(json) as TourGroupDetailsResponse

export const tourGroupDetailsResponseToJson = (value: TourGroupDetailsResponse): string =>
  JSON.stringify(value)

File: backend/src/main/scala/microservices/tour-group/objects/TourGroupDetailsResponse.scala
package com.typesafe.travel.tourgroup.domain

import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}

final case class TourGroupDetailsResponse(
    group: TourGroupSummaryResponse,
    memberships: List[TourGroupMembershipResponse],
    membershipTravelers: List[TourGroupMembershipTravelerResponse],
    planItems: List[GroupPlanItemResponse],
    planOptions: List[GroupPlanOptionResponse],
    selections: List[GroupPlanSelectionResponse],
    selectionOrderLinks: List[GroupSelectionOrderLinkResponse],
    blacklists: List[TourGroupBlacklistResponse],
    selectionOrderProjections: List[GroupSelectionOrderProjectionResponse] = Nil
)
object TourGroupDetailsResponse:
  given sourceEncoder: Encoder[TourGroupDetailsResponse] = deriveEncoder
  given sourceDecoder: Decoder[TourGroupDetailsResponse] = deriveDecoder
